src/ipips/ipip-0412.md

```diff
@@ -0,0 +1,240 @@
+   red and returned to the consumer) and when it's received might not match.
+1. Blocks can be reused multiple times, this is handy for cases when you plan
+   to cache on disk but not at all when you want to process a stream with use &
+   forget policy.
+
+What we really want is for the gateway to help us a bit, and give us blocks in
+a useful order.
+
+The existing Trustless Gateway specification does not provide a mechanism for
+negotiating the order of blocks in CAR responses.
+
+This IPIP aims to improve the status quo.
+
+## Detailed design
+
+CAR content type
+([`application/vnd.ipld.car`](https://www.iana.org/assignments/media-types/application/vnd.ipld.car))
+already supports `version` parameter, which allows gateway to indicate which
+CAR flavour is returned with the response.
+
+The proposed solution introduces two new parameters for the content type headers
+in HTTP requests and responses: `order` and `dups`.
+
+The `order` parameter allows the client to indicate its preference for a
+specific block order in the CAR response, and the `dups` parameter specifies
+whether duplicate blocks are allowed in the response.
+
+### Signaling in Request
+
+Content type negotiation is based on section 12.5.1 of :cite[rfc9110].
+
+Clients MAY indicate their preferred block order by sending an `Accept` header in
+the HTTP request. The `Accept` header format is as follows:
+
+```
+Accept: application/vnd.ipld.car; version=1; order=dfs; dups=y
+```
+
+In the future, when more orders or parameters exist, clients will be able to
+specify a list of preferences, for example:
+
+```
+Accept: application/vnd.ipld.car;order=foo, application/vnd.ipld.car;order=dfs;dups=y;q=0.5
+```
+
+The above example is a list of preferences, the client would really like to use
+the hypothetical `order=foo` however if this isn't available it would accept
+`order=dfs` with `dups=y` instead (lower priority indicated via `q` parameter,
+as noted in :cite[rfc9110]).
+
+#### `order` CAR content type parameter
+
+The `order` parameter accepts the following values:
+
+- `dfs`: [Depth-First Search](https://en.wikipedia.org/wiki/Depth-first_search)
+  order, allows for streaming responses with minimal memory usage
+- `rnd`: Unknown (random) order, the implicit default when `order` parameter is missing.
+
+#### `dups` CAR content type parameter
+
+The `dups` parameter specifies whether duplicate blocks (the same block
+occuring multiple times in the requested DAG) will be present in the CAR
+response.
+
+It accepts two values:
+- `y`: duplicate blocks are allowed
+- `n`: duplicates are not allowed
+
+When allowed (`y`), light clients are able to  discard blocks after
+reading them, removing the need for caching in-memory or on-disk.
+
+<!-- TODO: do we need a  parameter for inclusion of identity CIDs?
+It seems to be only relevant in Filecoin due to legacy hiccup:
+https://github.com/ipfs/specs/pull/330#issuecomment-1274106892 -->
```

Comment:
@rvagg seems at least @lidel thinks it's fair game to discuss here https://github.com/ipfs/specs/pull/412#issuecomment-1548773588.

Do you know of anyone that wants/needs this for exporting CARs using this API? I'm not personally aware of anyone and given that IIUC the trustless gateway components are being used mostly as data transports it doesn't seem like anyone should need identity CIDs in here. If no one needs this we should probably clarify that they will not be included.

Reply:
It's true! But we can change if needed.

Reply:
If we can strictly rule it out, at least in the first pass of this thing, then that'd be awesome. They cause so much pain and only made it into the original Filecoin CommP generation "spec" (really just two sides needing to agree and it being hard to signal a change after it was deployed) because we forgot to consider them, then spent many hours banging heads against walls making them properly handled.

They're unnecessary and wasteful, and a potential abuse vector too. Having it specifically spelled out in the spec would be great to make sure implementers all consider them.

Reply:
I'd like  to avoid them as well -- added sentence that clearly states they are never welcome (80c09d050b4776da4791829fb307de0a7b553bf4), lmk if we need more/less.

Reply:
Adding a checks to not include identity CIDs should be painless.